fendermint/vm/interpreter/src/fvm/gas/actor.rs

```diff
@@ -1,123 +1,109 @@
 use crate::fvm::gas::{Gas, GasMarket};
-use crate::fvm::state::FvmExecState;
 use crate::fvm::FvmMessage;
 use anyhow::Context;
 
-use fendermint_vm_actor_interface::gas::GAS_ACTOR_ADDR;
+use fendermint_vm_actor_interface::gas::GAS_MARKET_ACTOR_ADDR;
 use fendermint_vm_actor_interface::system;
-use fvm_ipld_blockstore::Blockstore;
+use fvm::executor::{ApplyKind, Executor};
 use fvm_ipld_encoding::BytesDe;
-use std::sync::atomic::{AtomicU64, Ordering};
+use fvm_shared::clock::ChainEpoch;
 
-type AtomicGas = AtomicU64;
 type GasMarketState = fendermint_actor_gas_market::EIP1559GasState;
 
-/// The gas market based on EIP1155
-/// Due to the reference trait bound limit (`&self` instead of `&mut self`) in Interpreter, `Atomic`
-/// is used. However, the calling pattern should be single threaded, so direct `store` could be used.
-/// The usage of `Atomic` is purely to bypass the compilation issue without using unsafe.
-/// TODO: remove this overhead when trait bound is updated.
 #[derive(Default)]
 pub struct ActorGasMarket {
     /// The block gas limit
-    block_gas_limit: AtomicGas,
+    block_gas_limit: Gas,
     /// The accumulated gas usage so far
-    block_gas_used: AtomicGas,
+    block_gas_used: Gas,
 }
 
 impl GasMarket for ActorGasMarket {
     fn available_block_gas(&self) -> Gas {
-        self.block_gas_limit.load(Ordering::SeqCst) - self.block_gas_used.load(Ordering::SeqCst)
+        self.block_gas_limit - self.block_gas_used
     }
 
-    fn consume_gas(&self, gas: Gas) -> anyhow::Result<()> {
-        let block_gas_used = self.block_gas_used.load(Ordering::SeqCst);
-
-        if block_gas_used + gas >= self.block_gas_limit.load(Ordering::SeqCst) {
+    fn record_gas_used(&mut self, gas: Gas) -> anyhow::Result<()> {
+        if self.block_gas_used + gas >= self.block_gas_limit {
             anyhow::bail!("out of block gas")
```

Comment:
This will induce a panic. This cannot happen; if it does, we have an internal assertion breakage. We really don't want to threaten the liveness of the chain with a panic. It's probably best to simply to a saturating add up to the block gas limit, and log a warning when we're exceeding it.

Reply:
works for me.